AudioKit/Common/Internals/AudioKit.swift

```diff
@@ -352,82 +355,76 @@ public typealias AKCallback = () -> Void
+        try start()
         if let playableNode = node as? AKToggleable {
             playableNode.play()
         }
         usleep(UInt32(duration * 1_000_000))
-        stop()
-        start()
+        try stop()
+        try start()
     }
 
     // MARK: - Configuration Change Response
 
     // Listen to changes in audio configuration
     // and restart the audio engine if it stops and should be playing
     @objc fileprivate static func restartEngineAfterConfigurationChange(_ notification: Notification) {
-        DispatchQueue.main.async {
+        do {
             if shouldBeRunning && !engine.isRunning {
-                do {
-
-                    #if !os(macOS)
-                        let appIsNotActive = UIApplication.shared.applicationState != .active
-                        let appDoesNotSupportBackgroundAudio = !AKSettings.appSupportsBackgroundAudio
-
-                        if appIsNotActive && appDoesNotSupportBackgroundAudio {
-                            AKLog("engine not restarted after configuration change since app was not active and does not support background audio")
-                            return
-                        }
-                    #endif
-
-                    try engine.start()
-                    dump(engine.outputNode.audioUnit)
+                #if !os(macOS)
+                    let appIsNotActive = UIApplication.shared.applicationState != .active
+                    let appDoesNotSupportBackgroundAudio = !AKSettings.appSupportsBackgroundAudio
 
-                    // Sends notification after restarting the engine, so it is safe to resume AudioKit functions.
-                    if AKSettings.notificationsEnabled {
-                        NotificationCenter.default.post(
-                            name: .AKEngineRestartedAfterConfigurationChange,
-                            object: nil,
-                            userInfo: notification.userInfo)
+                    if appIsNotActive && appDoesNotSupportBackgroundAudio {
+                        AKLog("engine not restarted after configuration change since app was not active and does not support background audio")
+                        return
                     }
+                #endif
 
-                } catch {
-                    AKLog("couldn't start engine after configuration change \(error)")
+                try engine.start()
+
+                // Sends notification after restarting the engine, so it is safe to resume AudioKit functions.
+                if AKSettings.notificationsEnabled {
+                    NotificationCenter.default.post(
+                        name: .AKEngineRestartedAfterConfigurationChange,
+                        object: nil,
+                        userInfo: notification.userInfo)
                 }
             }
+        } catch {
+            AKLog("error restarting engine after route change")
+            // Note: doesn't throw since this is called from a notification observer
```

Comment:
This is a code smell.  You mentioned thrashing as a motivation for this refactor.  Notifications that throw that can potentially trigger notifications (that also throw) seems like a good place to look for the thrashing.  Refactor if you have any ideas for how to deal with this, otherwise add a TODO, but it's definitely stinky :)

Reply:
Maybe there could be a callback that has a default value, engineWasInterrupted?  The default closure could restart the engine, but more serious apps would have an opportunity to control this behavior by setting a custom closure, then choosing whether or not to restart within the closure.